crates/crypto/merkle/src/multiproof.rs

```diff
@@ -0,0 +1,158 @@
+//! https://ethereum.github.io/consensus-specs/ssz/merkle-proofs/#merkle-multiproofs
+
+use std::collections::{BTreeMap, HashMap};
+
+use alloy_primitives::B256;
+use anyhow::ensure;
+use serde::{Deserialize, Serialize};
+
+use crate::{
+    hash_concat,
+    index::{
+        GeneralizedIndex, generalized_index_from_leaf_index, generalized_index_parent,
+        generalized_index_sibling, get_helper_indices,
+    },
+};
+
+/// Multiproof is a structure that contains the leaves to be verified with
+/// their corresponding proofs.
+#[derive(Debug, Clone, Default, Serialize, Deserialize)]
+pub struct Multiproof {
+    /// The leaves to be verified.
+    /// Keyed by their generalized indices.
+    pub leaves: HashMap<GeneralizedIndex, B256>,
+
+    /// The proof nodes.
+    /// Keyed by their generalized indices.
+    /// Keys of ``proofs`` will be sorted in descending order when generating a single proof.
+    pub proofs: BTreeMap<GeneralizedIndex, B256>,
+}
+
+impl Multiproof {
+    /// Generate a multiproof for the given tree and indices.
+    pub fn generate<const DEPTH: u64>(tree: &[B256], indices: &[u64]) -> anyhow::Result<Self> {
+        let bottom_length: u64 = 1 << DEPTH;
+
+        ensure!(!indices.is_empty(), "Indices cannot be empty");
+        for &index in indices {
+            ensure!(index < bottom_length, "Index out of bounds");
+        }
+
+        let generalized_indices: Vec<GeneralizedIndex> = indices
+            .iter()
+            .map(|&index| generalized_index_from_leaf_index(index, DEPTH))
+            .collect();
+        let helper_indices: Vec<GeneralizedIndex> = get_helper_indices(&generalized_indices);
+
+        let leaves: HashMap<GeneralizedIndex, B256> = generalized_indices
+            .iter()
+            .map(|&g| (g, tree[g as usize]))
+            .collect::<HashMap<_, _>>();
+        let proofs: BTreeMap<GeneralizedIndex, B256> = helper_indices
+            .iter()
+            .map(|&g| (g, tree[g as usize]))
+            .collect();
+
+        Ok(Self { leaves, proofs })
+    }
+
+    /// Return the root of the multiproof.
+    ///
+    /// Most code in this function is borrowed from ssz_rs crate.
+    /// https://github.com/ralexstokes/ssz-rs/blob/main/ssz-rs/src/merkleization/multiproofs.rs
```

Comment:
Could we just implement this function from the specs instead of borrowing code from ssz_rs, as I think it is better if we build independently.